ros2/ament_cmake_ros
# Add ament_add_pytest_isolated and ament_add_gtest_isolated

This PR depends on:
  - ~~https://github.com/ament/ament_cmake/pull/174~~ Already Merged

This PR blocks
  - https://github.com/ros2/launch/pull/256#issuecomment-499164065
  - https://github.com/ros2/ros_testing/pull/2

Use the new 'RUNNER' argument provided in ament_cmake PR 174 to create new versions of `ament_add_gtest` and `ament_add_pytest_test` that runs with ROS_DOMAIN_ID isolation if no ROS_DOMAIN_ID is already set.

This will also coordinate domain IDs with the ros2test tool

#### Testing
I've tested this PR in conjunction with ament_cmake 174 by changing rcl and rclpy to use the 'isolated' version of these macros, then running
```
terminal 1$ colcon test --packages-select rcl
terminal 2$ colcon test --packages-select rclpy
```
And verifying in the test logs that we see tests using ROS_DOMAIN_ID 1 as well as ROS_DOMAIN_ID 2

Test Cases:
  * Run without a ROS_DOMAIN_ID and verify that ROS_DOMAIN_ID is selected by looking at the latest test logs
  * Run to `colcon test` commands in parallel and verify that different ROS_DOMAIN_IDs are chosen
  * Set ROS_DOMAIN_ID and verify that it's not overridden in the log
  * Unset ROS_DOMAIN_ID and set DISABLE_ROS_ISOLATION and verify that ROS_DOMAIN_ID is not overridden in the log

#### Example usage
In rclpy:
```
--- a/ws/src/rclpy/rclpy/CMakeLists.txt                       
+++ b/ws/src/rclpy/rclpy/CMakeLists.txt                       
@@ -12,6 +12,7 @@ endif()                                                       
                                                                                
 find_package(ament_cmake REQUIRED)                                             
 find_package(ament_cmake_python REQUIRED)                                      
+find_package(ament_cmake_pytest_isolated REQUIRED)                             
 find_package(rcl REQUIRED)                                                     
 find_package(rcl_yaml_param_parser REQUIRED)                                   
 find_package(rcutils REQUIRED)                                                 
@@ -113,7 +114,7 @@ if(BUILD_TESTING)                                           
     string(REPLACE "\\" "/" ament_index_build_path "${ament_index_build_path}")
   endif()                                                                      
   if(NOT _typesupport_impls STREQUAL "")                                       
-    ament_add_pytest_test(rclpytests test                                      
+    ament_add_pytest_isolated_test(rclpytests test                             
       PYTHON_EXECUTABLE "${PYTHON_EXECUTABLE}"                                 
       APPEND_ENV AMENT_PREFIX_PATH=${ament_index_build_path}                   
         PYTHONPATH=${CMAKE_CURRENT_BINARY_DIR}                                 
```

In rcl:
```
--- a/ws/src/rcl/rcl/test/CMakeLists.txt                                                                                                  
+++ b/ws/src/rcl/rcl/test/CMakeLists.txt                                                                                                  
@@ -1,4 +1,5 @@                                                                                                                                             
 find_package(ament_cmake_gtest REQUIRED)                                                                                                                   
+find_package(ament_cmake_gtest_isolated REQUIRED)                                                                                                          
 find_package(ament_cmake_pytest REQUIRED)                                                                                                                  
                                                                                                                                                            
 find_package(test_msgs REQUIRED)                                                                                                                           
                                                                                                         
--- a/ws/src/rcl/rcl/test/cmake/rcl_add_custom_gtest.cmake                                                                                
+++ b/ws/src/rcl/rcl/test/cmake/rcl_add_custom_gtest.cmake                                                                                
@@ -74,7 +74,7 @@ macro(rcl_add_custom_gtest target)                                                                                                        
   endif()                                                                                                                                                  
                                                                                                                                                            
   # Pass args along to ament_add_gtest().                                                                                                                  
-  ament_add_gtest(${target} ${_ARG_SRCS} ${_ARG_ENV} ${_ARG_APPEND_ENV} ${_ARG_APPEND_LIBRARY_DIRS}                                                        
+  ament_add_gtest_isolated(${target} ${_ARG_SRCS} ${_ARG_ENV} ${_ARG_APPEND_ENV} ${_ARG_APPEND_LIBRARY_DIRS}                                               
                   ${_ARG_SKIP_TEST} ${_ARG_TIMEOUT})                                                                                                       
   # Check if the target was actually created.                                                                                                              
   if(TARGET ${target})                                                                                                                                     
```
